ipv8/messaging/anonymization/hidden_services.py

```diff
@@ -388,44 +410,67 @@ def on_linked_e2e(self, source_address, payload, circuit_id):
         else:
             self.logger.error('On linked e2e: could not find download for %s!', cache.info_hash)
 
-    def find_download(self, lookup_info_hash):
-        return self.service_callbacks.get(lookup_info_hash, None)
-
-    def create_introduction_point(self, info_hash, amount=1):
+    def create_introduction_point(self, info_hash, amount=1, required_ip=None):
         self.logger.info("Creating %d introduction point(s)", amount)
 
-        # Create a separate key per infohash
-        if info_hash not in self.session_keys:
-            self.session_keys[info_hash] = self.crypto.generate_key(u"curve25519")
+        if info_hash not in self.swarms:
+            self.logger.warning('Cannot create introduction point for unknown swarm')
+            return
+        elif not self.swarms[info_hash].seeding:
+            self.logger.warning('Cannot create introduction point for swarm that is not seeding')
+            return
+
+        seed_sk = self.swarms[info_hash].seeder_sk
 
         def callback(circuit):
             # We got a circuit, now let's create an introduction point
             circuit_id = circuit.circuit_id
-            self.my_intro_points[circuit_id].append(info_hash)
-
             cache = self.request_cache.add(IPRequestCache(self, circuit))
             self.send_cell([circuit.peer],
-                           u'establish-intro', EstablishIntroPayload(circuit_id, cache.number, info_hash))
+                           u'establish-intro', EstablishIntroPayload(circuit_id, cache.number, info_hash,
+                                                                     seed_sk.pub().key_to_bin()))
             self.logger.info("Established introduction tunnel %s", circuit_id)
 
         for _ in range(amount):
-            circuit_id = self.create_circuit_for_infohash(info_hash, CIRCUIT_TYPE_IP, callback)
-            self.infohash_ip_circuits[info_hash].append((circuit_id, time.time()))
+            self.create_circuit_for_infohash(info_hash, CIRCUIT_TYPE_IP_SEEDER, callback, required_exit=required_ip)
 
     @tc_lazy_wrapper_unsigned(EstablishIntroPayload)
     def on_establish_intro(self, source_address, payload, circuit_id):
+        if payload.public_key in self.intro_point_for:
+            self.logger.warning('Already have an introduction point for %s', binascii.hexlify(payload.public_key))
+            return
+
+        self.logger.info('Established introduction point for %s', binascii.hexlify(payload.public_key))
+
         circuit = self.exit_sockets[circuit_id]
-        self.intro_point_for[payload.info_hash] = circuit
+        self.intro_point_for[payload.public_key] = circuit, payload.info_hash
+
+        if not self.ipv8:
+            self.logger.error('No IPv8 service object available, cannot start PEXCommunity')
+        elif payload.info_hash not in self.pex:
+            community = PexCommunity(self.my_peer, self.ipv8.endpoint, self.ipv8.network, info_hash=payload.info_hash)
```

Comment:
You may want to use a new `Network()` here, to completely isolate the PEX peers from the other peers.

Reply:
Sure, I'll change this.